File: Assignment 5/Python/08. divide doubly linked list from middle and rejoin them in reverse manner .py
class Node:
    def __init__(self, val):
        self.prev = None
        self.val = val
        self.next = None

def traversal(head):
    ptr = head
    while ptr:
        print(ptr.val, end=" ")
        ptr = ptr.next
    print()

def divide_print(head):
    length = 0
    ptr = head
    while ptr:
        length += 1
        ptr = ptr.next

    ptr = head
    for _ in range(length // 2):
        ptr = ptr.next

    snd = ptr.next
    ptr.next = None
    snd.prev = None

    while snd.next:
        snd = snd.next

    snd.next = head
    head.prev = snd

    while ptr:
        print(ptr.val, end=" ")
        ptr = ptr.prev
    print()

if __name__ == "__main__":
    head = Node(1)
    snd = Node(2)
    trd = Node(3)
    four = Node(4)
    five = Node(5)
    six = Node(6)

    head.prev = None
    head.next = snd

    snd.prev = head
    snd.next = trd

    trd.prev = snd
    trd.next = four

    four.prev = trd
    four.next = five

    five.prev = four
    five.next = six

    six.prev = five
    six.next = None

    print("Original list: ", end="")
    traversal(head)
    print("Divided list: ", end="")
    divide_print(head)


# OUTPUT:
# Original list: 1 2 3 4 5 6 
# Divided list: 6 5 4 3 2 1 

File: Assignment 5/C/08. divide doubly linked list from middle and rejoin them in reverse manner.c
// input: 1->2->3->4->5->6
// Output:3->2->1->6->5->4

#include <stdio.h>
#include <stdlib.h>

typedef struct node{
    struct node* prev;
    int val;
    struct node* next;
}node;

void traversal(node* head){
    node* ptr = head;
    while(ptr){
        printf("%d ",ptr->val);
        ptr = ptr->next;
    }
    printf("\n");
}

void dividePrint(node* head){
    int s = 0;
    node* ptr = head;
    while(ptr){
        s++;
        ptr = ptr->next;
    }
    ptr = head;
    for(int i=1; i<s/2; i++){
        ptr = ptr->next;
    }
    node* snd = ptr->next;
    ptr->next = NULL;
    snd->prev = NULL;
    while(snd->next){
        snd = snd->next;
    }
    snd->next = head;
    head->prev = snd;

    while(ptr){
        printf("%d ",ptr->val);
        ptr = ptr->prev;
    }
    printf("\n");
}

int main(){
    node* head = (node*)malloc(sizeof(node));
    node* snd = (node*)malloc(sizeof(node));
    node* trd = (node*)malloc(sizeof(node));
    node* four = (node*)malloc(sizeof(node));

    head->val = 1;
    snd->val = 2;
    trd->val = 3;
    four->val = 4;

    head->prev = NULL;
    head->next = snd;

    snd->prev = head;
    snd->next = trd;
    
    trd->prev = snd;
    trd->next = four;

    four->prev = trd;
    four->next = NULL;

    traversal(head);
    dividePrint(head);

    return 0;
}
